rust/otel-arrow-rust/src/pdata/otlp/README.md

```diff
@@ -2,42 +2,246 @@
 
 Status: **Development**
 
-This package defines an OTLP pipeline data interface based on
-prost::Message objects.  The package layout:
-
-- `mod.rs`: Core module that re-exports the procedural macros and
-  defines the main OTLP message traits.
-- `tests.rs`: Contains test cases for the OTLP pipeline data
-  functionality.
-- `derive/`: A procedural macro crate that provides derive macros for
-  OTLP message types:
-  - `src/lib.rs`: Implements the `Message` procedural macro for
-    deriving OTLP interfaces.
-- `model/`: A crate that provides model definitions for OTLP types:
-  - `src/lib.rs`: Contains type definitions and metadata structures
-    for OTLP messages.
-
-## Design
-
 This package aims to resemble the OpenTelemetry Collector "pdata"
-interface, but for Rust.  Like the `pdatagen` command from that
-repository (Golang), this package includes a model definition that
-simplifies a few details, especially related to "oneof" features in
-the protocol.
+interface, but for Rust.  This package defines an OTLP pipeline data
+interface based on prost::Message objects
+
+For each OTLP message object, a set of additional implementation,
+associated methods and traits are generated through the ./derive
+sub-crate of this directory. 
+
+This package defines OTLP-specific structs relevant for encoding OTLP
+as bytes. It does this by implementing a Visitor pattern, in which:
+
+- Every OTLP message object has a corresponding Visitable trait, which is
+  the thing a Visitor will handle to visit the children of that message.
+- Every OTLP message object has a corresponding Visitor trait, which is how
+  to apply custom logic to OTLP-like data.
+- Every OTLP message object has also:
+  - MessageAdapter: a Visitable impl for OTLP message objects
+  - NoopVisitor: a do-nothing Visitor implements all traits
+  - Builder: a builder pattern for defning test OTLP message objects,
+    along with a new() method covering obligatory fields.
+  - EncodedLen: a Visitor for computing the sizes of an OTLP object
+    to an intermediate Vec<usize>, uses Accumulator, has test-only
+    pdata_size() implementation.
+  - Accumulator: an impl for summing the encoded size of the children
+    of an OTLP message.
 
 The file `model/src/lib.rs` defines how each message type is treated,
 with a few fields in each type being declared as parameters (in order
 of importance), with remaining fields set through a builder
 pattern. There are cases where the builder pattern is not used because
 all fields are specified as parameters.
 
-Oneof cases are expanded into one constructor per field. Since OTLP
-has no more than one oneof field per message, this leads to a simple
-pattern. The most important example is AnyValue, and in this case
-there are constructors named new_string, new_int, new_double, etc.
-
 See the [tests](./tests.rs) for examples of the resulting syntax.
 
-## Under development
+## Example generated code
```

Comment:
Everything in this readme is a great addition! super helpful to see the generated code as well as the tips for how to use `cargo expand` 👍